cmd/kcp-front-proxy/main.go

```diff
@@ -79,30 +83,58 @@ routed based on paths.`,
 				return errors.NewAggregate(errs)
 			}
 
-			var handler http.Handler
-			handler, err := proxy.NewHandler(&options.Proxy)
-			if err != nil {
-				return err
-			}
-
 			var servingInfo *genericapiserver.SecureServingInfo
 			var loopbackClientConfig *restclient.Config
 			if err := options.SecureServing.ApplyTo(&servingInfo, &loopbackClientConfig); err != nil {
 				return err
 			}
-
 			var authenticationInfo genericapiserver.AuthenticationInfo
 			if err := options.Authentication.ApplyTo(&authenticationInfo, servingInfo); err != nil {
 				return err
 			}
 
+			// start index
+			rootConfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(&clientcmd.ClientConfigLoadingRules{ExplicitPath: options.RootKubeconfig}, nil).ClientConfig()
+			if err != nil {
+				return fmt.Errorf("failed to load root kubeconfig: %w", err)
+			}
+			rootShardConfig, err := kcpclient.NewClusterForConfig(rootConfig)
+			if err != nil {
+				return fmt.Errorf("failed to create root client: %w", err)
+			}
+			kcpSharedInformerFactory := kcpinformers.NewSharedInformerFactoryWithOptions(rootShardConfig.Cluster(tenancyv1alpha1.RootCluster), 30*time.Minute)
+			indexController, err := index.NewController(
+				rootConfig.Host,
+				kcpSharedInformerFactory.Tenancy().V1alpha1().ClusterWorkspaceShards(),
+				func(shard *tenancyv1alpha1.ClusterWorkspaceShard) (kcpclient.ClusterInterface, error) {
+					shardConfig := restclient.CopyConfig(rootConfig)
+					shardConfig.Host = shard.Spec.BaseURL
+					shardClient, err := kcpclient.NewClusterForConfig(rootConfig)
```

Comment:
How come the root config is valid for each shard as well? Did we do some certificate magic?

Reply:
For now we assume some common CA. But that might change in the future.